Include govulncheck

adomaskizogian opened the pull request:
Updates `Dockerfile.template` with [govulncheck](https://pkg.go.dev/golang.org/x/vuln/cmd/govulncheck) installation steps.
Follows existing pattern of tools version management.

Addresses https://github.com/CircleCI-Public/cimg-go/issues/241

# Checklist

Please check through the following before opening your PR. Thank you!

- [x] I have made changes to the `Dockerfile.template` file only
- [x] I have not made any manual changes to automatically generated files
- [x] My PR follows best practices as described in the [contributing guidelines](CONTRIBUTING)
- [ ] (Optional, but recommended) My commits are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits)

BytesGuy commented:
@felicianotech Would appreciate a review from yourself on this one as you opened the feature request initially

BytesGuy reviewed (changes requested):
Thanks for the PR! A few changes requested :)

adomaskizogian commented on Dockerfile.template:
```diff
@@ -32,3 +32,12 @@ RUN curl -sSL "https://github.com/gotestyourself/gotestsum/releases/download/v${
+# by copying to output layer just the govulncheck binary
+FROM base AS govulncheck
+ENV GOVULNCHECK_V=1.0.0
+RUN go install "golang.org/x/vuln/cmd/govulncheck@v${GOVULNCHECK_V}"
```
I am not aware of how are these images maintained so naturally I just followed the pattern of how it is done with the other tools. I agree with your argument that it is an additional thing to take care especially that there is no need to have control of the version.

Removed pinned version, changed to Latest: https://github.com/CircleCI-Public/cimg-go/pull/242/commits/d403766d62f229faa1ce628194f2bf1ff563a638

adomaskizogian commented on Dockerfile.template:
```diff
@@ -32,3 +32,12 @@ RUN curl -sSL "https://github.com/gotestyourself/gotestsum/releases/download/v${
+
+# Install govulncheck on a temporary stage to ensure minimal size footprint
+# by copying to output layer just the govulncheck binary
+FROM base AS govulncheck
```
Reason why I've chosen to use multistage approach is to simplify the clean up. `go install` pulls dependencies, compiles them. This populates go mod and compiler build caches. This can be replaced with `go clean -cache -modcache && rm -rf "${GOPATH}/pkg"` but I am not 100% sure that this is the correct approach.

adomaskizogian replied:
Please let me know what you think of these 2 points

felicianotech replied:
I think it's better here to use `go clean -cache -modcache && rm -rf "${GOPATH}/pkg"` at the end and just install whatever is needed without a multistage build. This will make it easier if other `go install` based tools are also installed.

adomaskizogian replied:
@felicianotech addressed your remarks on https://github.com/CircleCI-Public/cimg-go/pull/242/commits/871a7192857023af1430ea47e9d007817d480a68

felicianotech commented on Dockerfile.template:
```diff
@@ -32,3 +32,11 @@ RUN curl -sSL "https://github.com/gotestyourself/gotestsum/releases/download/v${
+# Install govulncheck on a temporary stage to ensure minimal size footprint
+# by copying to output layer just the govulncheck binary
+FROM base AS govulncheck
+RUN go install "golang.org/x/vuln/cmd/govulncheck@latest"
```
This should be decided by @JalexChen I believe but I think we'd want to pin the version here and then add it to the updated script.

adomaskizogian replied:
Will address your comments by applying suggestions once JalexChen provides feedback on version pinning

adomaskizogian replied:
Addressed on https://github.com/CircleCI-Public/cimg-go/pull/242/commits/871a7192857023af1430ea47e9d007817d480a68

JalexChen replied:
i apologize -  i have not seen these tags for some reason, but i think this ended up being resolved in a later commit. i think the idea of having a version we can pin and is visible, rather than @latest is helpful for everyone if there needs to be debugging later on

adomaskizogian commented:
@BytesGuy which branch is the mainline? At the moment this PR targets main. I see there are fresher branches i.e. 1.20.6.

adomaskizogian commented:
@BytesGuy 
I've addressed review comments and sync with mainline branch on https://github.com/CircleCI-Public/cimg-go/pull/242/commits/871a7192857023af1430ea47e9d007817d480a68

BytesGuy commented on goFeed.sh:
```diff
@@ -9,6 +9,21 @@ else
   exit 1
 fi
 
+getGoVulnCheck() {
```
@JalexChen can you review this section as you are the most knowledgeable with regards to these scripts you set up previously, thanks :)

BytesGuy commented:
@adomaskizogian It's looking good, thanks :) Just would like @JalexChen to review the changes to the feed script

JalexChen reviewed (approved):
i think the changes here look great. the only thing that needs to be done now is to rebase on main as there have been a couple of changes.

- i ran this in a separate branch with it rebased and ran new builds as well. this should be good to merge. new builds should include this moving forward

adomaskizogian commented:
> i think the changes here look great. the only thing that needs to be done now is to rebase on main as there have been a couple of changes.
> 
> * i ran this in a separate branch with it rebased and ran new builds as well. this should be good to merge. new builds should include this moving forward

I've synced this PR with main.

adomaskizogian commented:
Thank you guys for accepting this pull request! Cant wait to start using it. Will look into the golang orb and include new job/command for govulncheck once I've got more free time on my hands.
